src/main/java/com/amazon/ion/impl/macro/PooledExpressionFactory.java

```diff
@@ -0,0 +1,380 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.ion.impl.macro;
+
+import com.amazon.ion.IonType;
+import com.amazon.ion.SymbolToken;
+import com.amazon.ion.Timestamp;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A factory for {@link Expression} instances. Avoids repetitive allocations by pooling instances. {@link #clear()}
```

Comment:
This is only used in the `EExpressionArgsReader` and related code, right? Should it be shared with the macro evaluator so that things like `make_string` can also use pooled `Expression` instances?

Reply:
Yes, that's a good callout. I'll work on that.